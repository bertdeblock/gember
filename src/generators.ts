import { generateDocument } from "./generate-document.js";
import { DocumentName } from "./types.js";

export function generateComponent(
  name: string,
  {
    authoringFormat = "gjs",
    classBased = false,
    cwd = "",
    path = "",
  }: {
    authoringFormat?: "gjs" | "gts";
    classBased?: boolean;
    cwd?: string;
    path?: string;
  } = {},
) {
  return generateDocument(DocumentName.Component, name, {
    cwd,
    inputs: { authoringFormat, classBased },
    path,
  });
}

export function generateHelper(
  name: string,
  {
    authoringFormat = "js",
    classBased = false,
    cwd = "",
    path = "",
  }: {
    authoringFormat?: "js" | "ts";
    classBased?: boolean;
    cwd?: string;
    path?: string;
  } = {},
) {
  return generateDocument(DocumentName.Helper, name, {
    cwd,
    inputs: { authoringFormat, classBased },
    path,
  });
}

export function generateModifier(
  name: string,
  {
    authoringFormat = "js",
    cwd = "",
    path = "",
  }: {
    authoringFormat?: "js" | "ts";
    cwd?: string;
    path?: string;
  } = {},
) {
  return generateDocument(DocumentName.Modifier, name, {
    cwd,
    inputs: { authoringFormat },
    path,
  });
}

export function generateService(
  name: string,
  {
    authoringFormat = "js",
    cwd = "",
    path = "",
  }: {
    authoringFormat?: "js" | "ts";
    cwd?: string;
    path?: string;
  } = {},
) {
  return generateDocument(DocumentName.Service, name, {
    cwd,
    inputs: { authoringFormat },
    path,
  });
}
