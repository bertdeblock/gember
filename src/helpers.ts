import type { EmberPackageJson } from "./types.js";

export function isV1Addon(packageJson: EmberPackageJson): boolean {
  if (isAddon(packageJson)) {
    const { version } = packageJson["ember-addon"] ?? {};

    return version === 1 || version === undefined;
  }

  return false;
}

export function isV2Addon(packageJson: EmberPackageJson): boolean {
  if (isAddon(packageJson)) {
    const { version } = packageJson["ember-addon"] ?? {};

    return version === 2;
  }

  return false;
}

function isAddon(packageJson: EmberPackageJson): boolean {
  if (Array.isArray(packageJson.keywords)) {
    return packageJson.keywords.includes("ember-addon");
  }

  return false;
}
