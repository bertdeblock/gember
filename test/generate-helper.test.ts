import fsExtra from "fs-extra";
import { readFile } from "node:fs/promises";
import { join } from "node:path";
import { afterEach, it } from "vitest";
import { generateHelper } from "../src/generators.ts";
import { copyBlueprint } from "./helpers.ts";

let cwd: string;

afterEach(() => fsExtra.remove(cwd));

it("generates a `.js` helper", async (ctx) => {
  cwd = await copyBlueprint("v2-addon");

  await generateHelper("foo", cwd);

  const content = await readFile(join(cwd, "src/helpers/foo.js"), "utf-8");

  ctx.expect(content).toMatchSnapshot();
});

it("generates a `.ts` helper", async (ctx) => {
  cwd = await copyBlueprint("v2-addon");

  await generateHelper("foo", cwd, { ts: true });

  const content = await readFile(join(cwd, "src/helpers/foo.ts"), "utf-8");

  ctx.expect(content).toMatchSnapshot();
});

it("generates a `.js` helper at a custom path", async (ctx) => {
  cwd = await copyBlueprint("v2-addon");

  await generateHelper("foo", cwd, { path: "src/-private" });

  const content = await readFile(join(cwd, "src/-private/foo.js"), "utf-8");

  ctx.expect(content).toMatchSnapshot();
});

it("generates a `.ts` helper at a custom path", async (ctx) => {
  cwd = await copyBlueprint("v2-addon");

  await generateHelper("foo", cwd, {
    path: "src/-private",
    ts: true,
  });

  const content = await readFile(join(cwd, "src/-private/foo.ts"), "utf-8");

  ctx.expect(content).toMatchSnapshot();
});
