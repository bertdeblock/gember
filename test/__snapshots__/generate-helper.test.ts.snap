// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generates a class-based \`.js\` helper 1`] = `
"import Helper from "@ember/component/helper";

export default class Foo extends Helper {
  compute(positional /*, named*/) {
    return positional;
  }
}
"
`;

exports[`generates a class-based \`.ts\` helper 1`] = `
"import Helper from "@ember/component/helper";

export default class Foo extends Helper {
  compute(positional /*, named*/) {
    return positional;
  }
}
"
`;

exports[`generates a function-based \`.js\` helper 1`] = `
"export default function foo(positional /*, named*/) {
  return positional;
}
"
`;

exports[`generates a function-based \`.js\` helper at a custom path 1`] = `
"export default function foo(positional /*, named*/) {
  return positional;
}
"
`;

exports[`generates a function-based \`.ts\` helper 1`] = `
"export default function foo(positional /*, named*/) {
  return positional;
}
"
`;

exports[`generates a function-based \`.ts\` helper at a custom path 1`] = `
"export default function foo(positional /*, named*/) {
  return positional;
}
"
`;
